File: app/src/main/java/ch/protonmail/android/jobs/ProtonMailBaseJob.java
/*
 * Copyright (c) 2020 Proton Technologies AG
 * 
 * This file is part of ProtonMail.
 * 
 * ProtonMail is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * ProtonMail is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with ProtonMail. If not, see https://www.gnu.org/licenses/.
 */
package ch.protonmail.android.jobs;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.birbit.android.jobqueue.Job;
import com.birbit.android.jobqueue.JobManager;
import com.birbit.android.jobqueue.Params;
import com.birbit.android.jobqueue.RetryConstraint;

import javax.inject.Inject;

import ch.protonmail.android.activities.messageDetails.repository.MessageDetailsRepository;
import ch.protonmail.android.api.ProtonMailApiManager;
import ch.protonmail.android.core.Constants;
import ch.protonmail.android.core.ProtonMailApplication;
import ch.protonmail.android.core.QueueNetworkUtil;
import ch.protonmail.android.core.UserManager;
import timber.log.Timber;

public abstract class ProtonMailBaseJob extends Job {

    @Inject
    protected transient UserManager mUserManager;
    @Inject
    protected transient ProtonMailApiManager mApi;
    @Inject
    protected transient JobManager mJobManager;
    @Inject
    protected transient QueueNetworkUtil mQueueNetworkUtil;
    @Inject
    protected transient MessageDetailsRepository messageDetailsRepository;
    protected String username;

    protected void inject() {
        ProtonMailApplication.getApplication().getAppComponent().inject(this);
    }

    protected ProtonMailBaseJob(Params params, String username) {
        super(params);
        inject();
        this.username = (username == null) ? mUserManager.getUsername() : username;
    }

    protected ProtonMailBaseJob(Params params) {
        this(params, null);
        inject();
    }

    @Override
    public void onAdded() {
    }

    @Override
    protected int getRetryLimit() {
        return Constants.JOB_RETRY_LIMIT_DEFAULT;
    }

    /**
     * Use {@link #onProtonCancel} for custom cancellation logic.
     */
    @Override
    protected final void onCancel(int cancelReason, @Nullable Throwable throwable) {
        Timber.e(throwable, this.getClass().getName() + " cancelled, reason = " + cancelReason + ", retryLimit = " + this.getRetryLimit());

        try {
            onProtonCancel(cancelReason, throwable);
        } catch (Exception e) {
            Timber.e(e, this.getClass().getName() + " threw exception in onProtonCancel");
            // throw e; this exception is swallowed by JobQueue anyway so there's no need to rethrow
        }
    }

    /**
     * Use this method for custom logic when Jobs get cancelled.
     */
    protected void onProtonCancel(int cancelReason, @Nullable Throwable throwable) {}

    @Override
    protected RetryConstraint shouldReRunOnThrowable(@NonNull Throwable throwable, int runCount, int maxRunCount) {
        return null;
    }
}

File: app/src/main/java/ch/protonmail/android/jobs/ProtonMailBaseJob.kt
/*
 * Copyright (c) 2020 Proton Technologies AG
 *
 * This file is part of ProtonMail.
 *
 * ProtonMail is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProtonMail is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ProtonMail. If not, see https://www.gnu.org/licenses/.
 */
package ch.protonmail.android.jobs

import ch.protonmail.android.core.Constants
import ch.protonmail.android.core.ProtonMailApplication
import ch.protonmail.android.di.JobEntryPoint
import com.birbit.android.jobqueue.Job
import com.birbit.android.jobqueue.Params
import com.birbit.android.jobqueue.RetryConstraint
import dagger.hilt.EntryPoints
import timber.log.Timber

abstract class ProtonMailBaseJob @JvmOverloads protected constructor(
    params: Params?,
    username: String? = null
) : Job(params) {

    protected val entryPoint get() =
        EntryPoints.get(ProtonMailApplication.getApplication(), JobEntryPoint::class.java)

    protected fun getApi() = entryPoint.apiManager()
    protected fun getJobManager() = entryPoint.jobManager()
    protected fun getMessageDetailsRepository() = entryPoint.messageDetailsRepository()
    protected fun getQueueNetworkUtil() = entryPoint.queueNetworkUtil()
    protected fun getUserManager() = entryPoint.userManager()

    @JvmField
    protected var username: String = username ?: getUserManager().username

    override fun onAdded() {}

    override fun getRetryLimit() = Constants.JOB_RETRY_LIMIT_DEFAULT

    /**
     * Use [.onProtonCancel] for custom cancellation logic.
     */
    override fun onCancel(cancelReason: Int, throwable: Throwable?) {
        Timber.d(throwable, "${javaClass.name} cancelled, reason = $cancelReason, retryLimit = $retryLimit")
        try {
            onProtonCancel(cancelReason, throwable)

        } catch (ignored: Exception) {
            Timber.e(ignored, this.javaClass.name + " threw exception in onProtonCancel")
        }
    }

    /**
     * Use this method for custom logic when Jobs get cancelled.
     */
    protected open fun onProtonCancel(cancelReason: Int, throwable: Throwable?) {}

    override fun shouldReRunOnThrowable(throwable: Throwable, runCount: Int, maxRunCount: Int): RetryConstraint? =
        null
}
